src/main.ml

```
@@ -83,68 +83,89 @@ let with_attached_vdi sr vdi read_write f =
+        Xen_api.Session.get_uuid ~rpc ~session_id:x ~self:x
+        >>= fun _ ->
+        return (x, false)
+      | Some uname, Some pwd, _ ->
+        Xen_api.Session.login_with_password ~rpc ~uname ~pwd ~version:"1.0" ~originator:"xapi-nbd"
+        >>= fun session_id ->
+        return (session_id, true)
+      | _, _, _ ->
+        fail (Failure "No suitable authentication provided")
+    ) >>= fun (session_id, need_to_logout) ->
+    Lwt.finalize
+      (fun () -> f session_id)
       (fun () ->
-        if need_to_logout
-        then Session.logout ~rpc ~session_id
-        else return ())
-  )
+         if need_to_logout
+         then Xen_api.Session.logout ~rpc ~session_id
+         else return ())
+  in
+
+  let channel = Nbd_lwt_unix.of_fd fd in
+  Lwt.finalize
+    (fun () ->
+       Nbd_lwt_unix.Server.connect channel ()
+       >>= fun (export_name, t) ->
+       Lwt.finalize
+         (fun () ->
+            let rpc = Xen_api.make !uri in
+            let uri = Uri.of_string export_name in
+            with_session rpc uri
+              (fun session_id ->
+                 let path = Uri.path uri in (* note preceeding / *)
+                 let vdi_uuid = if path <> "" then String.sub path 1 (String.length path - 1) else path in
+                 Xen_api.VDI.get_by_uuid ~rpc ~session_id ~uuid:vdi_uuid
+                 >>= fun vdi_ref ->
+                 Xen_api.VDI.get_record ~rpc ~session_id ~self:vdi_ref
+                 >>= fun vdi_rec ->
+                 Xen_api.SR.get_uuid ~rpc ~session_id ~self:vdi_rec.API.vDI_SR
+                 >>= fun sr_uuid ->
+                 with_attached_vdi sr_uuid vdi_rec.API.vDI_location (not vdi_rec.API.vDI_read_only)
+                   (fun filename ->
+                      with_block filename (Nbd_lwt_unix.Server.serve t (module Block))
+                   )
+              )
+         )
+         (fun () -> Nbd_lwt_unix.Server.close t)
+    )
+    (ignore_exn channel.close)
```

I'd add a comment here explaining why you need to ignore this exception